retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/StreamingCallSubscriber.java

```diff
@@ -0,0 +1,186 @@
+
+import java.io.IOException;
+import java.util.concurrent.Executor;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.google.common.base.Strings;
+
+import com.linecorp.armeria.client.retrofit2.ArmeriaCallFactory.ArmeriaCall;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpObject;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.HttpStatusClass;
+
+import okhttp3.Callback;
+import okhttp3.MediaType;
+import okhttp3.Protocol;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+import okio.Buffer;
+import okio.ForwardingSource;
+import okio.Okio;
+
+final class StreamingCallSubscriber implements Subscriber<HttpObject> {
+
+    private static final long NO_CONTENT_LENGTH = -1;
+
+    private final ArmeriaCall armeriaCall;
+    private final Callback callback;
+    private final Request request;
+    private final Executor callbackExecutor;
+    private final Response.Builder responseBuilder = new Response.Builder();
+    private final PipeBuffer pipeBuffer = new PipeBuffer();
+    @Nullable
+    private Subscription subscription;
+    private boolean nonInformationalHeadersStarted;
+    @Nullable
+    private String contentType;
+    private long contentLength = NO_CONTENT_LENGTH;
+    private boolean callbackCalled;
+
+    StreamingCallSubscriber(ArmeriaCall armeriaCall, Callback callback, Request request,
+                            Executor callbackExecutor) {
+        this.armeriaCall = armeriaCall;
+        this.callback = callback;
+        this.request = request;
+        this.callbackExecutor = callbackExecutor;
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        this.subscription = subscription;
+        if (armeriaCall.isCanceled()) {
+            onCancel(subscription);
+            return;
+        }
+        subscription.request(1);
+    }
+
+    @Override
+    public void onNext(HttpObject httpObject) {
+        // 'subscription' can never be null here because 'onNext' is invoked
+        // only after 'subscription' is set. See 'onSubscribe'.
+        assert subscription != null;
+        if (armeriaCall.isCanceled()) {
+            onCancel(subscription);
+            return;
+        }
+        if (httpObject instanceof HttpHeaders) {
+            subscription.request(1);
+            final HttpHeaders headers = (HttpHeaders) httpObject;
+            final HttpStatus status = headers.status();
+            if (!nonInformationalHeadersStarted) { // If not received a non-informational header yet
+                // Ignore informational headers or the headers without :status.
+                if (status == null || status.codeClass() == HttpStatusClass.INFORMATIONAL) {
```

Comment:
I see some code duplication between this class and `BlockingCallSubscriber`. Would it be possible to extract the common logic somewhere, like introducing a parent abstract class or a utility class for common logic?